Python/Product/Debugger/Debugger/PythonProcess.cs

```diff
@@ -1131,201 +907,176 @@ public PythonProcess(PythonLanguageVersion languageVersion, string exe, string a
+                language = unboundBreakpoint.IsDjangoBreakpoint ? LDP.LanguageKind.Django : LDP.LanguageKind.Python,
+                breakpointId = unboundBreakpoint.Id,
+                breakpointFileName = unboundBreakpoint.Filename,
+                breakpointLineNo = unboundBreakpoint.LineNo,
+            }, ct);
         }
 
-        internal void DisconnectRepl() {
-            DebugWriteCommand("Disconnect Repl");
-            lock (_streamLock) {
-                if (_stream == null) {
-                    return;
-                }
+        internal async Task ConnectReplAsync(int portNum, CancellationToken ct = default(CancellationToken)) {
+            await SendDebugRequestAsync(new LDP.ConnectReplRequest() {
+                portNum = portNum,
+            }, ct);
+        }
 
-                try {
-                    _stream.Write(DisconnectReplCommandBytes);
-                } catch (IOException) {
-                } catch (ObjectDisposedException) {
-                    // If the process has terminated, we expect an exception
-                }
+        internal async Task DisconnectReplAsync(CancellationToken ct = default(CancellationToken)) {
+            try {
+                await SendDebugRequestAsync(new LDP.DisconnectReplRequest() {
+                }, ct);
+            } catch (IOException) {
+            } catch (ObjectDisposedException) {
+                // If the process has terminated, we expect an exception
             }
         }
 
-        internal bool SetLineNumber(PythonStackFrame pythonStackFrame, int lineNo) {
+        internal async Task<bool> SetLineNumberAsync(PythonStackFrame pythonStackFrame, int lineNo, CancellationToken ct) {
             if (_stoppedForException) {
                 return false;
             }
 
-            DebugWriteCommand("Set Line Number");
-            lock (_streamLock) {
-                _setLineResult = false;
-                _stream.Write(SetLineNumberCommand);
-                _stream.WriteInt64(pythonStackFrame.Thread.Id);
-                _stream.WriteInt32(pythonStackFrame.FrameId);
-                _stream.WriteInt32(lineNo);
-            }
+            var timeoutSource = new CancellationTokenSource(2000);
+            var linkedSource = CancellationTokenSource.CreateLinkedTokenSource(ct, timeoutSource.Token);
 
-            // wait up to 2 seconds for line event...
-            for (int i = 0; i < 20 && _stream != null; i++) {
-                if (NativeMethods.WaitForSingleObject(_lineEvent.SafeWaitHandle, 100) == 0) {
-                    break;
-                }
-            }
+            var requestTask = SendDebugRequestAsync(new LDP.SetLineNumberRequest() {
+                threadId = pythonStackFrame.Thread.Id,
+                frameId = pythonStackFrame.FrameId,
+                lineNo = lineNo,
+            }, linkedSource.Token);
+
+            try {
+                return await requestTask.ContinueWith(t => {
```

Comment:
>await requestTask.ContinueWith [](start = 23, length = 30)

Why do we need to await/ContinueWith? If you want to avoid trying to marshal back to the original thread, use await requestTask.ConfigureAwait()